### DISCUSSION TITLE
Getting CharacterVirtual velocity aligned with ground plane

### DISCUSSION THREAD

**Discussion by ghost** [jrouwe/JoltPhysics#535] [General]
I'm trying to get a character setup in Jolt. I've got the basics copied from the sample but I'm stuck on one thing. I want my character to walk up slopes without the angle of the slope causing a divergence in the player input and the result. For example in my little test level my camera has my character pointing directly towards a specific spot on top of the hill:

![Screenshot 2023-05-17 21 16 55](https://github.com/jrouwe/JoltPhysics/assets/361789/6cde07b4-02ac-495f-8d5f-4fccc86d8866)

However if I hold W (without changing the camera or anything) the player ends up to the left of the desired location because the angle of the slope impacted the velocity:

![Screenshot 2023-05-17 21 17 01](https://github.com/jrouwe/JoltPhysics/assets/361789/7e3dd63f-57fa-4b39-9023-0c2940d428f1)

Is there a way to have the character align velocity with the ground as it moves? I've tried manipulating the movement input based on the camera and ground normal which I've had success with when writing a character controller from scratch in Unity, but it causes the player to pop off the ground at times, so I don't think this is quite correct for how Jolt's ExtendedUpdate works internally:

```c++
if (jph_move_amount.LengthSq() > 0)
{
    JPH::Quat move_rot = JPH::Quat::sEulerAngles({ 0.0f, camera->yaw, 0.0f });

    if (player->IsSupported() && (current_vertical_velocity.GetY() - ground_velocity.GetY()) < 0.1f)
    {
        move_rot = JPH::Quat::sFromTo({ 0, 1, 0 }, player->GetGroundNormal()) * move_rot;
    }

    jph_move_amount = move_rot * jph_move_amount;
}

new_velocity += jph_move_amount;
```

Is there another way? Maybe something in the character listener callbacks?

**Comment by jrouwe** [comment]
I think it would work better if you split up your movement in a component that runs parallel to the slope (i.e. where the movement is not affected by the slope) and a component that moves into the slope:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/42c3eed4-e29b-411e-9ec1-e3a3e6bb45ca)

In this plane, the movement (m) is projected by CharacterVirtual onto the slope and becomes shorter (p = m - m dot n * n). In order to move in the direction of the camera, you'd need to scale that movement component up by 1/cos(alpha) = 1/p dot m (or scale the parallel component down by cos(alpha)). Obviously this causes problems if the slope is vertical, so you'd need to handle that. 

When moving away from the slope you need a different scale vector as you will be moving faster rather than slower. 

The main advantage I see here is that you'd be moving horizontally only, which means that the character would not have the tendency to leave the floor.

Disclaimer: I didn't test the actual math so beware of mistakes. Note also that the projection mentioned above is a simplification of what the CharacterVirtual actually does, it can handle being in contact with more than 1 slope (imagine walking in a V) so a single ground normal cannot fully represent what the character is going to do (see: CharacterVirtual::SolveConstraints).

**Comment by ghost** [comment]
Interesting. I will take a look at that to see what I can incorporate but it sounds like what I want to achieve. Thanks!
